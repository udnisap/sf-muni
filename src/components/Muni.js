import React, { Component } from 'react';
import PropTypes from 'prop-types';
// eslint-disable-next-line no-unused-vars
import { transition } from 'd3-transition';
import { easeLinear } from 'd3-ease';
import { select } from 'd3-selection';
import { UPDATE_TIME, APPEARIN_TIME } from '../constants';

class Muni extends Component {
  componentWillEnter(cb) {
    select(this.node)
      .transition()
      .duration(APPEARIN_TIME)
      .ease(easeLinear)
      .style('fill-opacity', 1)
      .on('end', () => {
        this.setState({ alpha: 1 });
        cb();
      });
  }

  componentWillLeave(cb) {
    select(this.node)
      .transition()
      .duration(APPEARIN_TIME)
      .ease(easeLinear)
      .style('fill-opacity', 0)
      .on('end', () => {
        this.setState({ alpha: 0 });
        cb();
      });
  }

  componentWillReceiveProps(nextProps) {
    const { x, y } = nextProps;
    if (this.state.x !== x || this.state.y !== y) {
      select(this.node)
        .transition()
        .duration(UPDATE_TIME)
        .ease(easeLinear)
        .attr('cx', x)
        .attr('cy', y)
        .on('end', () => {
          this.setState({ x, y });
        });
    }
  }

  constructor(props) {
    super(props);
    const { x, y } = this.props;
    this.state = {
      alpha: 0,
      x,
      y
    };
  }

  render() {
    const { alpha, x, y } = this.state;
    return (
      <circle
        ref={node => (this.node = node)}
        cx={x}
        cy={y}
        r={5}
        className="muni"
        style={{ fillOpacity: alpha }}
      />
    );
  }
}

Muni.propTypes = {
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired
};

export default Muni;
