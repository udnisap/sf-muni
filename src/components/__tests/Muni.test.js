import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount, render } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
import Muni from '../Muni';

describe('Muni', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Muni x={2} y={3} />, div);
  });

  it('renders without crashing', () => {
    expect(shallow(<Muni x={2} y={3} />)).toMatchSnapshot();
  });
});
