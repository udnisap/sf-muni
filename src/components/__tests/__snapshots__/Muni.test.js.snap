// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Muni renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Muni
    x={2}
    y={3}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": undefined,
      "cx": 2,
      "cy": 3,
      "r": 5,
      "style": Object {
        "fillOpacity": 0,
      },
    },
    "ref": [Function],
    "rendered": null,
    "type": "circle",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": undefined,
        "cx": 2,
        "cy": 3,
        "r": 5,
        "style": Object {
          "fillOpacity": 0,
        },
      },
      "ref": [Function],
      "rendered": null,
      "type": "circle",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
